### PR TITLE
V2 generator

### PR DESCRIPTION
**Pull request by DevergneTimothee**
## Description
This PR aims to implement the new LITL loss for the infinitesimal generator. A softmax postprocessing layer can also be added in order to automatically learn the states. The constant function was also automatically added to the output vector during the training so that we do not need to learn it anymore

## Todos

  - [x]  Add proper tests
  - [x]  The forecast state occupation function can be made more general in order to compute the time evolution of experimentally relevant observable
  - [x] Add proper tutorial, (probably with muller three states)

## Questions
- [x] The learning can be divided in two distinct phases: 1) learn a representation, 2) learn the eigenfunctions. Shall we separate them more clearly, like TICA/deepTICA. This would maybe make things easier if we want to use a pre-trained representation

## Status
- [x] Tests need to be done

### CODE REVIEW

**Comment by EnricoTrizio** [comment]
Having two separated things for the representation and the eigenvalues could make sense. This is also the same situation we were discussing for SelfTICA (see #278). 
Isn't it @luigibonati ?

**Comment by DevergneTimothee** [comment]
> Having two separated things for the representation and the eigenvalues could make sense. This is also the same situation we were discussing for SelfTICA (see #278). Isn't it @luigibonati ?

I changed the structure of the code to match the one of the selfTICA, I added a folder in the core directory called "operator learning", where I added the Generator class which can be used without representation learning. This is just a proposition

**Comment by Kai-Zhu-2001** [comment]
Hi Timothee, 

I was thinking about the organization of the `Generator` class. Since it inherits from `Stats `and behaves similarly to `TICA`, `PCA`, and `LDA` — with a `compute()` method that estimates eigenvalues/eigenvectors and a `forward()` method that applies the projection — I think it may be more natural to place it under `mlcolvar/core/stats/.`

Maybe we could move the class to:

mlcolvar/core/stats/generator.py

and put the generator-specific helper functions, such as `compute_eigenfunctions` and `compute_covariances`, in a separate file:

mlcolvar/core/stats/generator_utils.py

I would probably keep `stats/utils.py` only for general statistical or linear algebra utilities, rather than mixing in the generator-specific graph/derivative/resolvent logic.

What do you think?

**Comment by DevergneTimothee** [comment]
> Hi Timothee,
> 
> I was thinking about the organization of the `Generator` class. Since it inherits from `Stats `and behaves similarly to `TICA`, `PCA`, and `LDA` — with a `compute()` method that estimates eigenvalues/eigenvectors and a `forward()` method that applies the projection — I think it may be more natural to place it under `mlcolvar/core/stats/.`
> 
> Maybe we could move the class to:
> 
> mlcolvar/core/stats/generator.py
> 
> and put the generator-specific helper functions, such as `compute_eigenfunctions` and `compute_covariances`, in a separate file:
> 
> mlcolvar/core/stats/generator_utils.py
> 
> I would probably keep `stats/utils.py` only for general statistical or linear algebra utilities, rather than mixing in the generator-specific graph/derivative/resolvent logic.
> 
> What do you think?

Indeed you are right, my bad. I created this "operator_learning" folder, because it is kind of a different object than the LDA and PCA, and would allow implementing other methods, like the reduced rank estimator for the transfer operator. But, either this deserves a separate baseclass, or we do as you said, which would be the easiest way

File: mlcolvar/cvs/generator/generator.py
Comment (Kai-Zhu-2001) [inline mlcolvar/cvs/generator/generator.py]: Since the `Generator` estimator already stores the eigenvectors and eigenvalues, do we need to duplicate them here? Could we remove these attributes and consistently use `self.generator.evecs` and `self.generator.evals` instead?

File: mlcolvar/cvs/generator/deep_generator.py [hunk 1]
```diff
@@ -106,123 +108,150 @@ def __init__(self,
...
+            if "activation" not in options[o]:
+                options[o]["activation"] = "tanh"
+        
+            self.nn = FeedForward(self.layers, **options[o])
+        else:
+            self.nn = model
+        
+        if self.softmax_postproc:
+            self.postprocessing=SoftmaxPostProcessing(r)
+        self.generator = Generator(in_features=r, out_features=r, feature_method=self.forward_nn)
```
Comment (Kai-Zhu-2001) [inline mlcolvar/cvs/generator/deep_generator.py:124]: The representation used here is inconsistent with the one used during training. training_step() applies the full postprocessing module, while the estimator receives the raw neural-network output and only applies softmax later, skipping the learnable linear transformation.

Could we define a shared method for the learned representation and use it consistently for both training and eigenfunction estimation?

File: mlcolvar/cvs/generator/deep_generator.py [hunk 2]
```diff
@@ -106,123 +108,150 @@ def __init__(self,
...
+    def forward_nn(self, x, cell=None):
+        if self.preprocessing is not None:
+            x = self._apply_module(self.preprocessing, x, cell=cell)
+        z = self.nn(x)
+        return z
+    def forward(self, x, cell=None):
+        if self.evecs is not None:
+            
+            output = self.forward_nn(x, cell=cell)
+            if self.softmax_postproc:
```
Comment (Kai-Zhu-2001) [inline mlcolvar/cvs/generator/deep_generator.py:204]: Training applies the full `SoftmaxPostProcessing` module, including the learnable linear transformation, whereas inference only applies softmax here. This means that training and final eigenfunction evaluation use different representations.

Could `forward()` reuse the same shared representation method used by training and the generator estimator? Also, if softmax is kept here, the dimension should be specified explicitly with `dim=-1`.

**Comment by Kai-Zhu-2001** [comment]
Hi Timothée, I think the macOS failure comes from the test comparing raw eigenvectors with fixed Linux reference values.

Eigenvectors from `torch.linalg.eigh` are only defined up to a sign, and nearly degenerate modes may also be rotated within the same subspace. Since macOS and Ubuntu use different linear algebra backends, both results can be correct while `torch.allclose` still fails.

I suggest removing the fixed `evecs` comparison, keeping a tolerant comparison of the eigenvalues, and comparing the eigenfunction subspaces through their projection matrices. A minimal alternative would be to align the eigenfunction signs before comparison, but the projector-based test would be more robust.

**Comment by DevergneTimothee** [comment]
> Hi Timothée, I think the macOS failure comes from the test comparing raw eigenvectors with fixed Linux reference values.
> 
> Eigenvectors from `torch.linalg.eigh` are only defined up to a sign, and nearly degenerate modes may also be rotated within the same subspace. Since macOS and Ubuntu use different linear algebra backends, both results can be correct while `torch.allclose` still fails.
> 
> I suggest removing the fixed `evecs` comparison, keeping a tolerant comparison of the eigenvalues, and comparing the eigenfunction subspaces through their projection matrices. A minimal alternative would be to align the eigenfunction signs before comparison, but the projector-based test would be more robust.

Thanks, the other option is to do like @EnricoTrizio did in the PR #284 what do you think?

**Comment by Kai-Zhu-2001** [comment]
Yes, I think it is a good opinion.
